File: modules/core/src/main/java/de/deutschebahn/bahnhoflive/ui/map/InitialPoiManager.java
package de.deutschebahn.bahnhoflive.ui.map;

import android.content.Intent;
import android.os.Bundle;
import android.os.Parcelable;

import androidx.annotation.Nullable;

public class InitialPoiManager {

    public static final String ARG_SOURCE = "initialPoiSource";
    public static final String ARG_POI = "initialPoi";
    public static final String STATE_DONE = "initialPoiManagerDone";

    private final Parcelable initialItem;

    @Nullable
    public final Content.Source source;

    private boolean done = false;

    public static void putInitialPoi(Intent intent, Content.Source source, Parcelable poiItem) {
        intent.putExtra(ARG_SOURCE, source);
        intent.putExtra(ARG_POI, poiItem);
    }

    public InitialPoiManager(Intent intent, Bundle savedInstanceState) {
        source = intent == null ? null : (Content.Source) intent.getSerializableExtra(ARG_SOURCE);

        if (savedInstanceState == null || !savedInstanceState.getBoolean(STATE_DONE, false)) {
            initialItem = intent.getParcelableExtra(ARG_POI);
        } else {
            initialItem = null;
        }
    }

    public void onSaveInstanceState(Bundle outState) {
        outState.putBoolean(STATE_DONE, done);
    }

    public boolean isInitial(MarkerBinder markerBinder) {
        return markerBinder.getMarkerContent().wraps(initialItem);
    }
}

File: modules/core/src/main/java/de/deutschebahn/bahnhoflive/ui/map/InitialPoiManager.kt
/*
 * SPDX-FileCopyrightText: 2020 DB Station&Service AG <bahnhoflive-opensource@deutschebahn.com>
 *
 * SPDX-License-Identifier: Apache-2.0
 */
package de.deutschebahn.bahnhoflive.ui.map

import android.content.Intent
import android.os.Bundle
import android.os.Parcelable
import de.deutschebahn.bahnhoflive.util.getParcelableExtraCompatible
import de.deutschebahn.bahnhoflive.util.getSerializableExtraCompatible

class InitialPoiManager(intent: Intent?, savedInstanceState: Bundle?) {

    private val initialItem: Parcelable?

    @JvmField
    val source: Content.Source?
    private val done = false

    init {
        source = intent?.getSerializableExtraCompatible(ARG_SOURCE, Content.Source::class.java)

        initialItem = intent?.takeIf {
            savedInstanceState == null || !savedInstanceState.getBoolean(STATE_DONE, false)
        }?.getParcelableExtraCompatible(ARG_POI, Parcelable::class.java)
    }

    fun onSaveInstanceState(outState: Bundle) {
        outState.putBoolean(STATE_DONE, done)
    }

    fun isInitial(markerBinder: MarkerBinder): Boolean {
        return markerBinder.markerContent.wraps(initialItem)
    }

    companion object {
        const val ARG_SOURCE = "initialPoiSource"
        const val ARG_POI = "initialPoi"
        const val STATE_DONE = "initialPoiManagerDone"

        @JvmStatic
        fun putInitialPoi(intent: Intent, source: Content.Source?, poiItem: Parcelable?) {
            intent.putExtra(ARG_SOURCE, source)
            intent.putExtra(ARG_POI, poiItem)
        }
    }
}
